flax/linen/attention.py

```diff
@@ -363,14 +250,13 @@ def __call__(self,
         query,
         key,
         value,
-        dtype=self.dtype,
-        axis=attention_axis,
         bias=attention_bias,
-        precision=self.precision,
         dropout_rng=dropout_rng,
         dropout_rate=self.dropout_rate,
         broadcast_dropout=self.broadcast_dropout,
-        deterministic=self.deterministic)
+        deterministic=self.deterministic,
+        dtype=self.dtype,
+        precision=self.precision)
 
     # back to the original inputs dimensions
     out = DenseGeneral(features=features,
```

Comment:
now that things are efficient, clean and simple, don't we want to also get rid of  `DenseGeneral`s as well?

Reply:
Hmm, I think the use of DenseGeneral here and above is nice, actually.  What you often do otherwise is define two "split_heads", "join_heads" util functions that do the two reshapes... and DenseGeneral is sort of just a nice concise way to do that.  Since we used it in mlperf I also know that it performs fine.

Reply:
Totally agree that `DenseGeneral` is quite nice.  I was just referring to a tiny bit of improvement we can get from removing it, which might not be true given the extra reshape/transpose stuff we add with "split_heads", "join_heads" (which also makes the code a bit ugly).

p.s.
Just for the record, I profiled these two cases (which is wat is happening in DenseGeneral as well, the flattening applying dot_general on on axis was slightly better than applying dot_general on multiple axis:

```pyhton
rng = random.PRNGKey(0)
rng1, rng2 = random.split(rng)

inputs_shape = (8, 32, 32, 16, 16)  # bs, h, w, heads, features
inputs = random.normal(rng1, inputs_shape)
kernel = random.normal(rng2, (1, 16, 16))

inputs_flatten = inputs.reshape((8, 32, 32, 256))
kernel_flatten = kernel.reshape((1, 256))

@jax.jit
def dot_2d(inputs, kernel):
  return jax.lax.dot_general(inputs,  kernel,
                          (((3, 4), (1, 2)), ((), ())),
                          precision=jax.lax.Precision.HIGHEST)
@jax.jit
def dot_flatten(inputs, kernel):
  return jax.lax.dot_general(inputs,  kernel,
                          (((3,), (1,)), ((), ())),
                          precision=jax.lax.Precision.HIGHEST)

```